x-pack/filebeat/input/http_endpoint/input.go

```diff
@@ -0,0 +1,290 @@
+
+	"github.com/elastic/beats/v7/filebeat/channel"
+	"github.com/elastic/beats/v7/filebeat/input"
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+const (
+	inputName = "http_endpoint"
+)
+
+func init() {
+	err := input.Register(inputName, NewInput)
+	if err != nil {
+		panic(errors.Wrapf(err, "failed to register %v input", inputName))
+	}
+}
+
+type HttpEndpoint struct {
+	config
+	log      *logp.Logger
+	outlet   channel.Outleter // Output of received messages.
+	inputCtx context.Context  // Wraps the Done channel from parent input.Context.
+
+	workerCtx    context.Context         // Worker goroutine context. It's cancelled when the input stops or the worker exits.
+	workerCancel context.CancelFunc      // Used to signal that the worker should stop.
+	workerOnce   sync.Once               // Guarantees that the worker goroutine is only started once.
+	workerWg     sync.WaitGroup          // Waits on worker goroutine.
+	server       *HttpServer             // Server instance
+	httpRequest  http.Request            // Current Request
+	httpResponse http.ResponseWriter     // Current ResponseWriter
+	eventObject  *map[string]interface{} // Current event object
+}
+
+// NewInput creates a new httpjson input
+func NewInput(
+	cfg *common.Config,
+	connector channel.Connector,
+	inputContext input.Context,
+) (input.Input, error) {
+	// Extract and validate the input's configuration.
+	conf := defaultConfig()
+	if err := cfg.Unpack(&conf); err != nil {
+		return nil, err
+	}
+
+	// Build outlet for events.
+	out, err := connector.ConnectWith(cfg, beat.ClientConfig{
+		Processing: beat.ProcessingConfig{
+			DynamicFields: inputContext.DynamicFields,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap input.Context's Done channel with a context.Context. This goroutine
+	// stops with the parent closes the Done channel.
+	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
+	go func() {
+		defer cancelInputCtx()
+		select {
+		case <-inputContext.Done:
+		case <-inputCtx.Done():
+		}
+	}()
+
+	// If the input ever needs to be made restartable, then context would need
+	// to be recreated with each restart.
+	workerCtx, workerCancel := context.WithCancel(inputCtx)
+
+	in := &HttpEndpoint{
+		config:       conf,
+		log:          logp.NewLogger(inputName),
+		outlet:       out,
+		inputCtx:     inputCtx,
+		workerCtx:    workerCtx,
+		workerCancel: workerCancel,
+	}
+
+	// Create an instance of the HTTP server with the beat context
+	in.server, err = createServer(in)
+	if err != nil {
+		return nil, err
+	}
+
+	in.log.Infof("Initialized %v input on %v:%v", inputName, in.config.ListenAddress, in.config.ListenPort)
+	return in, nil
+}
+
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *HttpEndpoint) Run() {
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
```

Comment:
You have to call `in.workerWg.Done()` from the function `in.run`. The goroutine you start is executed "in paralel", it return immediately. Thus, `in.workerWg.Done()` is called right after the start. To wait until the execution of `in.run` is done, you need to rewrite the code.

```golang
func (in *HttpEndpoint) Run() {
	in.workerOnce.Do(func() {
		in.workerWg.Add(1)
		go in.run()
	})
}

func (in *HttpEndpoint) run() {
	defer in.workerWg.Done()
	defer in.log.Infof("%v worker has stopped.", inputName)
	err := in.server.Start()
	if err != nil {
		return
	}

	return
}
```